domain/src/main/java/org/oppia/domain/topic/TopicController.kt

```diff
@@ -191,6 +245,215 @@ class TopicController @Inject constructor(
+        if (storyProgress.chapterProgressMap.containsKey(chapterSummary.explorationId)) {
+          val chapterBuilder = chapterSummary.toBuilder()
+          chapterBuilder.chapterPlayState = ChapterPlayState.COMPLETED
+          storyBuilder.setChapter(chapterIndex, chapterBuilder)
+        } else {
+          if (storyBuilder.getChapter(chapterIndex - 1).chapterPlayState == ChapterPlayState.COMPLETED) {
+            val chapterBuilder = chapterSummary.toBuilder()
+            chapterBuilder.chapterPlayState = ChapterPlayState.NOT_STARTED
+            storyBuilder.setChapter(chapterIndex, chapterBuilder)
+          } else {
+            val chapterBuilder = chapterSummary.toBuilder()
+            chapterBuilder.chapterPlayState = ChapterPlayState.NOT_PLAYABLE_MISSING_PREREQUISITES
+            storyBuilder.setChapter(chapterIndex, chapterBuilder)
+          }
+        }
+      }
+      return storyBuilder.build()
+    } else {
+      return setFirstChapterAsNotStarted(storySummary)
+    }
+  }
+
+  // TODO(#21): Expose this as a data provider, or omit if it's not needed.
+  internal fun retrieveTopic(topicId: String): Topic {
+    return when (topicId) {
+      TEST_TOPIC_ID_0 -> createTestTopic0()
+      TEST_TOPIC_ID_1 -> createTestTopic1()
+      FRACTIONS_TOPIC_ID -> createTopicFromJson(
+        "fractions_topic.json", "fractions_skills.json", "fractions_stories.json"
+      )
+      RATIOS_TOPIC_ID -> createTopicFromJson(
+        "ratios_topic.json", "ratios_skills.json", "ratios_stories.json"
+      )
+      else -> throw IllegalArgumentException("Invalid topic ID: $topicId")
+    }
+  }
+
+  private fun retrieveTopicAsync(topicId: String): AsyncResult<Topic> {
+    return when (topicId) {
+      TEST_TOPIC_ID_0 -> AsyncResult.success(createTestTopic0())
+      TEST_TOPIC_ID_1 -> AsyncResult.success(createTestTopic1())
+      FRACTIONS_TOPIC_ID -> AsyncResult.success(
+        createTopicFromJson(
+          "fractions_topic.json", "fractions_skills.json", "fractions_stories.json"
+        )
+      )
+      RATIOS_TOPIC_ID -> AsyncResult.success(
+        createTopicFromJson(
+          "ratios_topic.json", "ratios_skills.json", "ratios_stories.json"
+        )
+      )
+      else -> AsyncResult.failed(IllegalArgumentException("Invalid topic ID: $topicId"))
+    }
+  }
+
+  private fun retrieveStory(storyId: String): StorySummary {
+    return when (storyId) {
+      TEST_STORY_ID_0 -> createTestTopic0Story0()
+      TEST_STORY_ID_1 -> createTestTopic0Story1()
+      TEST_STORY_ID_2 -> createTestTopic1Story2()
+      FRACTIONS_STORY_ID_0 -> createStoryFromJsonFile("fractions_stories.json", /* index= */ 0)
+      RATIOS_STORY_ID_0 -> createStoryFromJsonFile("ratios_stories.json", /* index= */ 0)
+      RATIOS_STORY_ID_1 -> createStoryFromJsonFile("ratios_stories.json", /* index= */ 1)
+      else -> StorySummary.getDefaultInstance()
+    }
+  }
+
+  private fun retrieveStoryAsync(storyId: String): AsyncResult<StorySummary> {
```

Comment:
Ditto for this method.

Reply:
Okay, Actually I was not sure that it will turn into AsyncResult.failed() or not, but now as I run test cases, it makes sure that it does work therefore removing both of these Async functions.